cmd/tpm2_autounlock/tpm2_autounlock.sh

```diff
@@ -0,0 +1,176 @@
+rootguid=$(zfs get guid -o value -H "${ENCRYPTIONROOT}")
+
+# Use -i or --index value if specified, otherwise look for existing property, and finally fallback to script default
+if [ -z "$tpmindex" ]; then
+	tpmindex=$(zfs get -H -o value org.zfsonlinux.tpm2:index "${ENCRYPTIONROOT}")
+	if [ "$tpmindex" = "-" ]; then
+		tpmindex="$defaulttpmindex"
+	fi
+fi
+
+# Use -P or --pcrs value if specified, otherwise look for existing property, and finally fallback to script default
+if [ -z "$tpmpcrs" ]; then
+	tpmpcrs=$(zfs get -H -o value org.zfsonlinux.tpm2:pcrs "${ENCRYPTIONROOT}")
+	if [ "$tpmpcrs" = "-" ]; then
+		tpmpcrs="$defaulttpmpcrs"
+	fi
+fi
+
+# If password is needed, use argument, piped, "script default" or stdin, in that order
+if [ -z "$password" ] && (( verifyonly+clearonly+nvlockonly == 0)); then
+	[[ -p /dev/stdin ]] && { mapfile -t; set -- "${MAPFILE[@]}"; set -- $@; }
+	pipepassword="$@"
+	if [ -n "$pipepassword" ]; then
+		password="$pipepassword"
+	elif [ -n "$defaultpassword" ]; then
+		password="$defaultpassword"
+	else
+		while true; do
+			read -s -p "Drive unlock password: " password
+			echo
+   			read -s -p "Confirm Password: " password2
+			echo
+  			[ "$password" = "$password2" ] && break
+    			echo "Please try again"
+		done
+	fi	
+fi
+
+# Check for invalid arguments
+if (( verifyonly+clearonly+nvlockonly > 1)) ; then
+	echo " You can only select only one of the following: verifyonly, clearonly, or nvlockonly"
+	exit 1
+fi
+
+########################################
+# Start execution
+if [ $clearonly ]; then
+	tpm2_nvundefine "$tpmindex" >/dev/null 2>&1
+	zfs inherit org.zfsonlinux.tpm2:index "${ENCRYPTIONROOT}"
+	zfs inherit org.zfsonlinux.tpm2:pcrs "${ENCRYPTIONROOT}"
+
+elif [ $verifyonly ]; then
+	echo "Attempting to read TPM index '$tpmindex' locked with pcrs '$tpmpcrs' and guid '$rootguid'."
+	define_policy="--policy"
+	cmd="tpm2_nvread ${tpmindex} --auth=session:s.dat+${rootguid}"
+	tpm_session_execute
+	if [ $? = 0 ]; then
+		echo "is the stored password."
+#	else
+		# The tpm2_nvread stdout messages will be visible to user
+	fi
+
+elif [ $nvlockonly ]; then
+	define_policy="--policy"
+	cmd="tpm2_nvreadlock  ${tpmindex} --auth=session:s.dat+${rootguid}"
+	tpm_session_execute
+else
+	# Preemptively clear the tpm index
+	tpm2_nvundefine "$tpmindex" >/dev/null 2>&1
+
+	# Define NVRAM location and its access rules
+	cmd="tpm2_nvdefine ${tpmindex} -Q --hierarchy=o --index-auth=${rootguid} --size=512 --policy=policy.dat --attributes='policyread|policywrite|read_stclear'"
+	tpm_session_execute
+
+	# Write password to NVRAM location
+	define_policy="--policy"
+	cmd="echo $password | tpm2_nvwrite ${tpmindex} -Q --auth=session:s.dat+${rootguid} --input=-"
+	tpm_session_execute
+	echo "Note: Password storage can be verified with '--verifyonly' until reboot or locked manually with '--nvlockonly'."
+
+	# Store zfs filesystem properties
+	zfs set org.zfsonlinux.tpm2:index="$tpmindex" "${ENCRYPTIONROOT}"
```

Comment:
since you store the index, you might want to search for an open NV slot and use that rather than requiring a hard coded default or user to specify.
```bash
tpm2_getcap handles-nv-index
```
Will give you the current in use handles...

Reply:
That is true.  Additionally, looking at the 4.1 changelog: 
"tpm2_nvdefine: Support searching for free index if an index isn't specified."  
If I bump the required version to 4.1+, I may be able to have tpm2_nvdefine do the work for me.

Reply:
Filed https://github.com/tpm2-software/tpm2-tools/issues/1891
I'd like to use tpm2_nvdefine to search for a usable index, but it offered an index not appropriate to the owner hierarchy as you showed in your spec document above.

Reply:
> That is true. Additionally, looking at the 4.1 changelog:
> "tpm2_nvdefine: Support searching for free index if an index isn't specified."
> If I bump the required version to 4.1+, I may be able to have tpm2_nvdefine do the work for me.

Yeah it just starts at index 0, perhaps we should smarten that up.

Reply:
> I'd like to use tpm2_nvdefine to search for a usable index, but it offered an index not appropriate to the owner hierarchy as you showed in your spec document above.

Maybe add an option --search=<stop index> which makes the argument the starting index and searches until stop is reached.